datadog_lambda/tracing.py

```diff
@@ -170,23 +185,98 @@ def extract_context_from_http_event_or_context(event, lambda_context):
     return trace_id, parent_id, sampling_priority
 
 
-def extract_context_from_sqs_event_or_context(event, lambda_context):
+def create_sns_event(message):
+    return {
+        "Records": [
+            {
+                "EventSource": "aws:sns",
+                "EventVersion": "1.0",
+                "Sns": message,
+            }
+        ]
+    }
+
+
+def extract_context_from_sqs_or_sns_event_or_context(event, lambda_context):
     """
     Extract Datadog trace context from the first SQS message attributes.
 
     Falls back to lambda context if no trace data is found in the SQS message attributes.
     """
     try:
         first_record = event["Records"][0]
-        msg_attributes = first_record.get("messageAttributes", {})
-        dd_json_data = msg_attributes.get("_datadog", {}).get("stringValue", r"{}")
+
+        # logic to deal with SNS => SQS event
+        if "body" in first_record:
+            body_str = first_record.get("body", {})
+            try:
+                body = json.loads(body_str)
+                if body.get("Type", "") == "Notification" and "TopicArn" in body:
+                    logger.debug("Found SNS message inside SQS event")
+                    first_record = get_first_record(create_sns_event(body))
+            except Exception:
+                first_record = event["Records"][0]
+                pass
+
+        msg_attributes = first_record.get(
+            "messageAttributes",
+            first_record.get("Sns", {}).get("MessageAttributes", {}),
+        )
+        dd_payload = msg_attributes.get("_datadog", {})
+        dd_json_data = dd_payload.get("stringValue", dd_payload.get("Value", r"{}"))
         dd_data = json.loads(dd_json_data)
         trace_id = dd_data.get(TraceHeader.TRACE_ID)
         parent_id = dd_data.get(TraceHeader.PARENT_ID)
         sampling_priority = dd_data.get(TraceHeader.SAMPLING_PRIORITY)
 
         return trace_id, parent_id, sampling_priority
-    except Exception:
+    except Exception as e:
+        logger.debug("The trace extractor returned with error %s", e)
+        return extract_context_from_lambda_context(lambda_context)
+
+
+def extract_context_from_eventbridge_event(event, lambda_context):
+    """
+    Extract datadog trace context from an EventBridge message's Details.
+    Details is often a weirdly escaped almost-JSON string. Here we have to correct for that.
```

Comment:
This comment says that you need to extract the context from an almost-JSON like string, and that you have to correct for it, but I don't see anything in the function about extracting the context from a string or correcting any formatting --  am I missing something?

Reply:
Yeah, I think Chris added that comment because of a bug in moto where `Details` is a JSON string that's not properly escaped & requires using something like `ast.literal_eval` rather than `json.loads()`. However, this is just an issue with the testing library, and not an actually in AWS, so I've updated the comment.